### PR TITLE
8296904: Improve handling of macos xcode toolchain

### PR DESCRIPTION
**Pull request by RealCLanger**
With this PR I'd like to make it easier to use dedicated installations of Xcode on Mac OS for OpenJDK builds without having to switch the active Xcode globally via `xcode-select`.

I propose adding a new configure flag `--with-xcode-path` that takes the path to an Xcode installation. If the option is set, this path is expanded to a valid `TOOLCHAIN_PATH`.

Furthermore, I fix detection of xcodebuild and correctly setting the sysroot from the toolchain by moving `AC_SUBST(TOOLCHAIN_PATH)` before calling `BASIC_SETUP_XCODE_SYSROOT` and honoring `TOOLCHAIN_PATH` in `BASIC_SETUP_XCODE_SYSROOT` in the case when no devkit is specified. As I see it, this is a viable fix, even if not introducing `--with-xcode-path`.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8296904](https://bugs.openjdk.org/browse/JDK-8296904): Improve handling of macos xcode toolchain

### Reviewers
 * [Erik Joelsson](https://openjdk.org/census#erikj) (@erikj79 - **Reviewer**) ⚠️ Review applies to [dd808218](https://git.openjdk.org/jdk/pull/11113/files/dd808218fd7c8a6d0edcb73b4b154fa5b35531a3)
 * [Magnus Ihse Bursie](https://openjdk.org/census#ihse) (@magicus - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11113/head:pull/11113` \
`$ git checkout pull/11113`

Update a local copy of the PR: \
`$ git checkout pull/11113` \
`$ git pull https://git.openjdk.org/jdk pull/11113/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11113`

View PR using the GUI difftool: \
`$ git pr show -t 11113`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11113.diff">https://git.openjdk.org/jdk/pull/11113.diff</a>

</details>

### CODE REVIEW

File: make/autoconf/basic.m4 [hunk 1]
```diff
@@ -293,6 +302,22 @@ AC_DEFUN_ONCE([BASIC_SETUP_DEVKIT],
       [UTIL_PREPEND_TO_PATH([TOOLCHAIN_PATH],$with_toolchain_path)]
   )
 
+  AC_ARG_WITH([xcode-path], [AS_HELP_STRING([--with-xcode-path],
+      [set up toolchain on Mac OS using a path to an xcode installation])])
```
Comment (erikj79) [inline make/autoconf/basic.m4:306]:
```suggestion
      [set up toolchain on Mac OS using a path to an Xcode installation])])
```

**Review by erikj79 — APPROVED** [review]
I think this looks good. It's technically possible to achieve what you need using the existing `--with-toolchain-path`, but in the case of Xcode, that is rather cumbersome as you need to point to the internal `bin` directories in the Xcode installation instead of just the top level Xcode application dir.

**Comment by RealCLanger** [comment]
Thanks for the review, I took over your suggestion. Will integrate shortly.

**Review by magicus — APPROVED** [review]
Looks good, thanks for fixing this!

**Comment by magicus** [comment]
Is there perhaps some documentation in doc/building.md that needs updating? I know there is a bit written about Xcode, I can imagine some pointers to this new argument would fit in there.

**Comment by RealCLanger** [comment]
> Is there perhaps some documentation in doc/building.md that needs updating? I know there is a bit written about Xcode, I can imagine some pointers to this new argument would fit in there.

Good suggestion. I made some edits in building.md. Please have a look. Also, what about building.html? Is this generated and if so, how can I regenerate it?

**Comment by magicus** [comment]
It is generated, and you can re-generate it using `make update-build-docs`. However, you need to have the same version of pandoc installed and configured, otherwise you'll get a lot of spurious changes. 

To help you out, I checked out your branch and regenerated it for you. You can find it at https://github.com/openjdk/jdk/commit/66ef4b730656daeb4108f7118b92dea3e8da6475 (from https://github.com/magicus/jdk/tree/xcode-toolchain). If you are not a git wrangling pro, you can download it as a diff and apply it locally by appending `.diff`: https://github.com/openjdk/jdk/commit/66ef4b730656daeb4108f7118b92dea3e8da6475.diff

**Comment by RealCLanger** [comment]
> It is generated, and you can re-generate it using `make update-build-docs`. However, you need to have the same version of pandoc installed and configured, otherwise you'll get a lot of spurious changes.
> 
> To help you out, I checked out your branch and regenerated it for you. You can find it at [66ef4b7](https://github.com/openjdk/jdk/commit/66ef4b730656daeb4108f7118b92dea3e8da6475) (from https://github.com/magicus/jdk/tree/xcode-toolchain). If you are not a git wrangling pro, you can download it as a diff and apply it locally by appending `.diff`: https://github.com/openjdk/jdk/commit/66ef4b730656daeb4108f7118b92dea3e8da6475.diff

Woho, [66ef4b7] is a super power trick 😄 I've added it and then figured out that my new enumeration isn't displayed well in html. Can you please regnerate it once more?

Other than that, are my edits ok?

**Comment by magicus** [comment]
Yes, I thought your edits were okay. I'll re-generated the html for you, and then I think you're good to go.

**Comment by magicus** [comment]
Hm, the changes you made to the markdown file does not result in any changes in the generated html file. :( I fixed it by adding an additional empty line in the markdown file. The commit including both these fixes are here: https://github.com/magicus/jdk/commit/2844de1591ee8246dfb0fb5d19664b9db1d68c3a

**Comment by magicus** [comment]
@RealCLanger Finally regenerated once more: [fcd20c73ed9f0a44c0614a54580c755a16387a6a](https://github.com/magicus/jdk/commit/fcd20c73ed9f0a44c0614a54580c755a16387a6a)

**Review by magicus — APPROVED** [review]

**Comment by RealCLanger** [comment]
Thanks @magicus for your help.

/integrate

**Pull request closed without merging**